File: SourceCode/Python/ChalmersGU/OpenAddressingHashMap.py

from API import Map

#/* *** ODSATag: OpenAddressingHashMap *** */
#/* *** ODSATag: Header *** */
class OpenAddressingHashMap(Map):
#/* *** ODSAendTag: Header *** */
#/* *** ODSATag: Constants *** */
    _minCapacity = 8
    _minLoadFactor = 0.3
    _maxLoadFactor = 0.7
    _capacityMultiplier = 1.5
#/* *** ODSAendTag: Constants *** */
#/* *** ODSATag: Header *** */

    def __init__(self):
        self._initialiseTable(self._minCapacity)

    def _initialiseTable(self, capacity):
        self._internalTable = [None] * capacity
        self._mapSize = 0
        self._numDeleted = 0
#/* *** ODSAendTag: Header *** */

#/* *** ODSATag: Probe *** */
    def _probe(self, key, i):
        return i       # Linear probing
        # return i*i   # Quadratic probing
#/* *** ODSAendTag: Probe *** */

#/* *** ODSATag: HashIndex *** */
    def _hashAndProbe(self, key):
        home = hash(key) & 0x7fffffff   # Turn the hash code into a positive integer
        for i in range(len(self._internalTable)):
            pos = (home + self._probe(key, i)) % len(self._internalTable)
            elem = self._internalTable[pos]
            if elem is None or key == elem.key:
                return pos
        raise ValueError("Hash table full")
#/* *** ODSAendTag: HashIndex *** */

#/* *** ODSATag: Put *** */
    def put(self, key, value):
        i = self._hashAndProbe(key)
        elem = self._internalTable[i]
        old = None
        if elem is None:
            self._internalTable[i] = KVPair(key, value)
            self._mapSize += 1
        else:
            old = elem.value
            elem.value = value
        if self.loadFactor() > self._maxLoadFactor:
            self._resizeTable(len(self._internalTable) * self._capacityMultiplier)
        return old
#/* *** ODSAendTag: Put *** */

#/* *** ODSATag: Get *** */
    def get(self, key):
        i = self._hashAndProbe(key)
        elem = self._internalTable[i]
        return None if elem is None else elem.value
#/* *** ODSAendTag: Get *** */

#/* *** ODSATag: Remove *** */
    def remove(self, key):
        i = self._hashAndProbe(key)
        elem = self._internalTable[i]
        if elem is None:
            return None
        removed = elem.value
        elem.key = None
        elem.value = None
        self._mapSize -= 1
        self._numDeleted += 1
        if self._mapSize < self._minLoadFactor * len(self._internalTable):
            self._resizeTable(len(self._internalTable) / self._capacityMultiplier);
        return removed
#/* *** ODSAendTag: Remove *** */

    def containsKey(self, key):
        i = self._probe(key)
        return self._internalTable[i] is not None

#/* *** ODSATag: Resize *** */
    def _resizeTable(self, newCapacity):
        if newCapacity < self._minCapacity: return
        oldTable = self._internalTable
        self._initialiseTable(int(newCapacity))
        for elem in oldTable:
            if elem is not None and elem.key is not None:
                self.put(elem.key, elem.value)
#/* *** ODSAendTag: Resize *** */

    def isEmpty(self):
        return self._mapSize == 0

    def size(self):
        return self._mapSize

#/* *** ODSATag: LoadFactor *** */
    def loadFactor(self):
        return (self._mapSize + self._numDeleted) / len(self._internalTable)
#/* *** ODSAendTag: LoadFactor *** */

    def __iter__(self):
        for elem in self._internalTable:
            if elem is not None and elem.key is not None:
                yield elem.key


#/* *** ODSATag: KVPair *** */
# Python does not have internal classes, so we have to make the k/v pairs standalone.
class KVPair:
    def __init__(self, key, value):
        self.key = key
        self.value = value
#/* *** ODSAendTag: KVPair *** */
#/* *** ODSAendTag: OpenAddressingHashMap *** */


#######################################################################################
## What comes below is purely for debugging and testing purposes - it can be removed ##

def _printMap(m):
    print(len(m._internalTable), "[",
          " ".join("- " if e is None else "# " if e.key is None else f"{e.key}:{e.value}" for e in m._internalTable),
          "]", m.size(), "{", m.loadFactor(), "}")

if __name__ == '__main__':
    m = OpenAddressingHashMap()
    _printMap(m)
    print("Putting values")
    for i in range(40):
        m.put(chr((i%25)+65), i)
        if i % 5 == 0: _printMap(m)
    _printMap(m)
    print("Removing values")
    i = 7
    while not m.isEmpty():
        m.remove(chr((i%25)+65))
        i += 11
        if m.size() % 5 == 0: _printMap(m)

File: SourceCode/Java_Generic/ChalmersGU/API/OpenAddressingHashMap.java

import java.util.Iterator;

/* *** ODSATag: OpenAddressingHashMap *** */
/* *** ODSATag: Header *** */
class OpenAddressingHashMap<K, V> implements Map<K, V> {
    KVPair<K,V>[] internalTable;
    int mapSize;
    int numDeleted;

/* *** ODSAendTag: Header *** */
/* *** ODSATag: Constants *** */
    static int MinCapacity = 8;
    static double MinLoadFactor = 0.3;
    static double MaxLoadFactor = 0.7;
    static double CapacityMultiplier = 1.5;
/* *** ODSAendTag: Constants *** */

/* *** ODSATag: Header *** */
    OpenAddressingHashMap() {
        initialiseTable(MinCapacity);
    }

    @SuppressWarnings("unchecked")
    private void initialiseTable(int capacity) {
        internalTable = (KVPair<K,V>[]) new KVPair[capacity];
        mapSize = 0;
        numDeleted = 0;
    }
/* *** ODSAendTag: Header *** */

/* *** ODSATag: HashIndex *** */
    private int hashAndProbe(K key) {
        int home = key.hashCode() & 0x7fffffff;   // Turn the hash code into a positive integer
        for (int i = 0; i < internalTable.length; i++) {
            int pos = (home + probe(key, i)) % internalTable.length;
            KVPair<K,V> elem = internalTable[pos];
            if (elem == null || key.equals(elem.key))
                return pos;
        }
        throw new IllegalStateException("Hash table full");
    }

    private int probe(K key, int i) {
        return i;        // Linear probing
        // return i*i;   // Quadratic probing
    }
/* *** ODSAendTag: HashIndex *** */

/* *** ODSATag: Put *** */
    public V put(K key, V value) {
        int i = hashAndProbe(key);
        KVPair<K,V> elem = internalTable[i];
        V old = null;
        if (elem == null) {
            internalTable[i] = new KVPair<>(key, value);
            mapSize++;
        } else {
            old = elem.value;
            elem.value = value;
        }
        if (loadFactor() > MaxLoadFactor)
            resizeTable((int) (internalTable.length * CapacityMultiplier));
        return old;
    }
/* *** ODSAendTag: Put *** */

/* *** ODSATag: Get *** */
    public V get(K key) {
        int i = hashAndProbe(key);
        KVPair<K,V> elem = internalTable[i];
        return elem == null ? null : elem.value;
    }
/* *** ODSAendTag: Get *** */

/* *** ODSATag: Remove *** */
    public V remove(K key) {
        int i = hashAndProbe(key);
        KVPair<K,V> elem = internalTable[i];
        if (elem == null)
            return null;
        V removed = elem.value;
        elem.key = null;
        elem.value = null;
        mapSize--;
        numDeleted++;
        if (mapSize < MinLoadFactor * internalTable.length) 
            resizeTable((int) (internalTable.length / CapacityMultiplier));
        return removed;
    }            
/* *** ODSAendTag: Remove *** */

    public boolean containsKey(K key) {
        int i = hashAndProbe(key);
        return internalTable[i] != null;
    }

/* *** ODSATag: Resize *** */
    private void resizeTable(int newCapacity) {
        if (newCapacity < MinCapacity) return;
        KVPair<K,V>[] oldTable = internalTable;
        initialiseTable(newCapacity);
        for (int i = 0; i < oldTable.length; i++) {
            KVPair<K,V> elem = oldTable[i];
            if (elem != null && elem.key != null)
                put(elem.key, elem.value);
        }
    }
/* *** ODSAendTag: Resize *** */

    public boolean isEmpty() {
        return mapSize == 0;
    }

    public int size() {
        return mapSize;
    }

/* *** ODSATag: LoadFactor *** */
    public double loadFactor() {
        return (double) (mapSize + numDeleted) / internalTable.length;
    }
/* *** ODSAendTag: LoadFactor *** */

    public Iterator<K> iterator() {
        return new HashMapIterator();
    }

    private class HashMapIterator implements Iterator<K> {
        private int index;
        HashMapIterator() {
            index = 0;
        }
        public boolean hasNext() {
            KVPair<K,V> elem = internalTable[index];
            while (elem == null || elem.key == null) {
                index++;
                if (index >= mapSize)
                    return false;
                elem = internalTable[index];
            }
            return true;
        }
        public K next() {
            hasNext();
            KVPair<K,V> elem = internalTable[index];
            index++;
            return elem.key;
        }
    }
/* *** ODSAendTag: OpenAddressingHashMap *** */


/***************************************************************************************/
/** What comes below is purely for debugging and testing purposes - it can be removed **/

    public void _printMap() {
        System.out.print(internalTable.length + " [ ");
        for (int i=0; i<internalTable.length; i++) {
            KVPair<K,V> elem = internalTable[i];
            System.out.print(elem == null ? "- " : elem.key == null ? "# " : elem.key + ":" + elem.value + " ");
        }
        System.out.println("] " + size() + " {" + loadFactor() + "}");
    }

    public static void main(String[] args) {
        OpenAddressingHashMap<String, Integer> map = new OpenAddressingHashMap<>();
        map._printMap();
        System.out.println("Putting values");
        for (int i=0; i<40; i++) {
            map.put(String.valueOf((char)((i%25)+65)), i);
            if (i % 5 == 0) map._printMap();
        }
        map._printMap();
        System.out.println("Removing values");
        for (int i = 7; !map.isEmpty(); i += 11) {
            map.remove(String.valueOf((char)((i%25)+65)));
            if (map.size() % 3 == 0) map._printMap();
        }
    }
/* *** ODSATag: OpenAddressingHashMap *** */
}
/* *** ODSAendTag: OpenAddressingHashMap *** */

